container-search/src/main/java/com/yahoo/search/dispatch/Dispatcher.java

```diff
@@ -57,16 +58,39 @@ public class Dispatcher extends AbstractComponent {
     private final DispatchConfig dispatchConfig;
     private final RpcResourcePool rpcResourcePool;
     private final SearchCluster searchCluster;
-    private final ClusterMonitor<Node> clusterMonitor;
     private volatile VolatileItems volatileItems;
 
     private static class VolatileItems {
+
         final LoadBalancer loadBalancer;
         final InvokerFactory invokerFactory;
-        VolatileItems(LoadBalancer loadBalancer, InvokerFactory invokerFactory) {
+        final ClusterMonitor<Node> clusterMonitor;
+        final AtomicInteger inflight = new AtomicInteger(1); // Initial reference.
+        Runnable cleanup = () -> { };
+
+        VolatileItems(LoadBalancer loadBalancer, InvokerFactory invokerFactory, ClusterMonitor<Node> clusterMonitor) {
             this.loadBalancer = loadBalancer;
             this.invokerFactory = invokerFactory;
+            this.clusterMonitor = clusterMonitor;
+        }
+
+        private void countDown() {
+            if (inflight.decrementAndGet() == 0) cleanup.run();
+        }
+
+        private class Ref implements AutoCloseable {
```

Comment:
Consider using resouce/reference concepts from jdisc_core. Take a look at `OnnxRuntime` and `ReferencedOrtSession`

Reply:
Yes, I just wanted this to be as simple as possible, but I'll take a look at that and see if it's better.